src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -368,89 +370,109 @@ ErrorIfQueryNotSupported(Query *queryTree)
+	const char *errorHint = NULL;
+	const char *joinHint = "Consider joining tables on partition column and have "
+						   "equal filter on joining columns.";
+	const char *filterHint = "Consider using an equality filter on the distributed "
+							 "table's partition column.";
 
 	if (queryTree->hasSubLinks)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Subqueries other than in from-clause are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasWindowFuncs)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Window functions are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->setOperations)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Union, Intersect, or Except are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasRecursive)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Recursive queries are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->cteList)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Common Table Expressions are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasForUpdate)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "For Update/Share commands are currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->distinctClause)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Distinct clause is currently unsupported";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->groupingSets)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Grouping sets, cube, and rollup is currently unsupported";
+		errorHint = filterHint;
 	}
 
 	hasTablesample = HasTablesample(queryTree);
 	if (hasTablesample)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Tablesample is currently unsupported";
+		errorHint = filterHint;
 	}
 
 	hasUnsupportedJoin = HasUnsupportedJoinWalker((Node *) queryTree->jointree, NULL);
 	if (hasUnsupportedJoin)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Join types other than inner/outer joins are currently unsupported";
+		errorHint = joinHint;
 	}
 
 	hasComplexJoinOrder = HasComplexJoinOrder(queryTree);
 	if (hasComplexJoinOrder)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Complex join orders are currently unsupported";
+		errorHint = joinHint;
 	}
 
 	hasComplexRangeTableType = HasComplexRangeTableType(queryTree);
 	if (hasComplexRangeTableType)
 	{
 		preconditionsSatisfied = false;
 		errorDetail = "Complex table expressions are currently unsupported";
+		errorHint = filterHint;
 	}
 
+
 	/* finally check and error out if not satisfied */
 	if (!preconditionsSatisfied)
 	{
+		bool showHint = ErrorHintRequired(errorHint, queryTree);
 		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 						errmsg("cannot perform distributed planning on this query"),
-						errdetail("%s", errorDetail)));
+						errdetail("%s", errorDetail),
+						showHint ? errhint("%s", errorHint) : 0));
```

Comment:
@jasonmp85 (or @mtuncer) -- just wanted to add a small note.

Inlining a ternary operator within a macro definition reads a bit complex to me. What if we simply had an if statement with two ereport()s?